icefoganalytics/travel-authorization
# Refactor: Normalize Component Pattern To Better Handle Per-Tab Stateful Access

Fixes https://github.com/icefoganalytics/travel-authorization/issues/137

Relates to:
- https://github.com/icefoganalytics/travel-authorization/pull/136

# Context

**Is your feature request related to a problem? Please describe.**
With the current component pattern, components are broken out at the page level by action.
Examples are:

- Edit
- Read
- Manage

As it turns out, I need to add a Edit tab to a Read layout which doesn't make sense with the current pattern. In this case I need to add an Expense tab that shows up when the Details and Estimates tabs are readonly.

Also need to permit update permissions for expenses when travel authorization is approved.

**Describe the solution you'd like**
Concerns:

I don't want to mix and match edit/read/manage code on the same page as I know that kind of pattern is _very_ hard to maintain and read.

Current URL

http://localhost:8080/my-travel-requests/4/edit/expense

I might be able to fix my component pattern with a URL like
http://localhost:8080/my-travel-requests/4/expense/edit
i.e. each tab would have a separate edit/read state.

I'm not sure how to handle manage ... but could defer that for now?

New pattern
-> components should be as generic as possible
-> components should accept ids or objects
-> components should avoid being current user specific, if possible, this should take effect at the lowest level, rather than the highest as I'm currently doing

path: "/my-travel-requests",
-> MyTravelAuthorizationsPage.vue

path: "/my-travel-requests/:travelAuthorizationId",
-> MyTravelAuthorizationLayout.vue
    -> sub-components per "user action?" or at least components that are as generic as possible
    -> should redirect to details page
path: "/my-travel-requests/:travelAuthorizationId/details",
    -> MyTravelAuthorizationDetailsPage.vue
        -> redirect to /details/edit if editable or stay here if read only
        -> components should be as generic as possible
path: "/my-travel-requests/:travelAuthorizationId/estimate",
    -> MyTravelAuthorizationEstimatePage.vue
        -> redirect to /estimate/edit if editable or stay here if read only
path: "/my-travel-requests/:travelAuthorizationId/expense",
    -> MyTravelAuthorizationExpensePage.vue
        -> redirect to /expense/edit if editable or stay here if read only
path: "/my-travel-requests/:travelAuthorizationId/details/edit",
path: "/my-travel-requests/:travelAuthorizationId/estimate/edit",
path: "/my-travel-requests/:travelAuthorizationId/expense/edit",

path: "/manager-view",
-> ManageTravelAuthorizationsPage.vue

path: "/travel-requests/:travelAuthorizationId",
-> TravelAuthorizationPage.vue
    -> sub-components are as generic as possible, possibly use the same components as the /my-travel-requests/:travelAuthorizationId pages?
    -> should redirect to details page
path: "/travel-requests/:travelAuthorizationId/details/manage",
path: "/travel-requests/:travelAuthorizationId/estimate/manage",

# Implementation

Implement per-tab stateful urls. i.e. each tab queries the vuex store and determines whether to show the link for the edit or read page for that tab.

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
2. Boot the app via `dev up`
3. Log in to the app at http://localhost:8080
4. Go to the "My Travel Requests" page via the top drop down nav.
5. Create a new travel authorization via the "+Travel Authorization" button.
6. Check that the Expense tab is disabled, and has a tooltip explaining why.
7. Fill in the Details panel and then scroll down to the Approvals panel.
8. Click the Generate Estimate button, then save and generate.
9. Submit the travel authorization to any email but yours, then click "Submit to Supervisor".
10. Use your database manager to manually approve the travel request.
11. You will now see that expenses tab is enabled.
    > Note that you are now in the "Edit" view again, but the tabs for Details and Estimates still point to the read views, as of this PR.
12. Click on the Prefill Expenses button to prefill your expenses
    > Note that you have have policy access to edit Expenses, even while Estimate editing is locked, as of this PR.
13. Check that you can edit expenses.
14. Check that you can delete all your expenses via the red "x" button, and prefill them again.
15. Check that you can add receipts and that once you do, the button changes to a "view receipt" button.
16. Check that you can download receipts via the view receipt button.
